physx/src/traits/descriptor.rs

```diff
@@ -0,0 +1,395 @@
+}
+
+pub struct SceneDescriptor<
+    U,
+    L: ArticulationLink,
+    S: RigidStatic,
+    D: RigidDynamic,
+    T: Articulation,
+    C: ArticulationReducedCoordinate,
+    OC: CollisionCallback,
+    OT: TriggerCallback,
+    OCB: ConstraintBreakCallback,
+    OWS: WakeSleepCallback<L, S, D>,
+    OA: AdvanceCallback<L, D>,
+> {
+    pub phantom_marker: PhantomData<(L, S, D, T, C)>,
+    pub user_data: U,
+    pub on_collide: Option<OC>,
+    pub on_trigger: Option<OT>,
+    pub on_constraint_break: Option<OCB>,
+    pub on_wake_sleep: Option<OWS>,
+    pub on_advance: Option<OA>,
+    pub gravity: PxVec3,
+    pub kine_kine_filtering_mode: PairFilteringMode,
+    pub static_kine_filtering_mode: PairFilteringMode,
+    pub broad_phase_type: BroadPhaseType,
+    pub limits: SceneLimits,
+    pub friction_type: FrictionType,
+    pub solver_type: SolverType,
+    pub bounce_threshold_velocity: f32,
+    pub friction_offset_threshold: f32,
+    pub ccd_max_separation: f32,
+    pub solver_offset_slop: f32,
+    pub flags: BitFlags<SceneFlag>,
+    pub static_structure: PruningStructureType,
+    pub dynamic_structure: PruningStructureType,
+    pub dynamic_tree_rebuild_rate_hint: u32,
+    pub scene_query_update_mode: SceneQueryUpdateMode,
+    pub solver_batch_size: u32,
+    pub solver_articulation_batch_size: u32,
+    pub nb_contact_data_blocks: u32,
+    pub max_nb_contact_data_blocks: u32,
+    pub max_bias_coefficient: f32,
+    pub contact_report_stream_buffer_size: u32,
+    pub ccd_max_passes: u32,
+    pub ccd_threshold: f32,
+    pub wake_counter_reset_value: f32,
+    pub sanity_bounds: PxBounds3,
+
+    pub simulation_filter_shader: FilterShaderDescriptor,
+
+    pub thread_count: u32, //pub cpu_dispatcher: *mut PxCpuDispatcher,
+    pub broad_phase_callback: *mut physx_sys::PxBroadPhaseCallback,
+    pub contact_modify_callback: *mut physx_sys::PxContactModifyCallback,
+    pub ccd_contact_modify_callback: *mut physx_sys::PxCCDContactModifyCallback,
+
+    pub cuda_context_manager: *mut physx_sys::PxCudaContextManager,
+    pub gpu_dynamics_config: physx_sys::PxgDynamicsMemoryConfig,
+    pub gpu_max_num_partitions: u32,
+    pub gpu_compute_version: u32,
+}
+
+impl<
+        U,
+        L: ArticulationLink,
+        S: RigidStatic,
+        D: RigidDynamic,
+        T: Articulation,
+        C: ArticulationReducedCoordinate,
+        OC: CollisionCallback,
+        OT: TriggerCallback,
+        OCB: ConstraintBreakCallback,
+        OWS: WakeSleepCallback<L, S, D>,
+        OA: AdvanceCallback<L, D>,
+    > SceneDescriptor<U, L, S, D, T, C, OC, OT, OCB, OWS, OA>
+{
+    pub fn new(user_data: U) -> Self {
+        Self {
+            phantom_marker: PhantomData,
+            user_data,
+            on_collide: None,
+            on_trigger: None,
+            on_constraint_break: None,
+            on_wake_sleep: None,
+            on_advance: None,
+            gravity: Default::default(),
```

Comment:
Could we write all of these out instead of using Default::default()?   In case of gravity here, this will be zero, right? I think we should default gravity to 0, -9.81, 0, for example.

Reply:
We should use the default PhysX values, in my opinion. Especially as some engines use Z-up, and some don't want gravity.

Reply:
Oh, hm. I thought that was the default, but of course, zero is the right value here.

Anyway, for bools and enums, I don't really like to see Default::default().

Reply:
I replaced all the defaults, and changed gravity to be -9.81 y.  PhysX default is 0g 🤷

Reply:
I don't think we should change gravity from default; that seems like a gotcha.

Reply:
Oops, shoulda refreshed before I commited